libqtile/backend/wayland/core.py

```diff
@@ -0,0 +1,346 @@
+        )
+        self.qtile.process_button_motion(self.cursor.x, self.cursor.y)
+
+    def _add_new_pointer(self, device: input_device.InputDevice):
+        logger.info("Adding new pointer")
+        self.cursor.attach_input_device(device)
+
+    def _add_new_keyboard(self, device: input_device.InputDevice):
+        logger.info("Adding new keyboard")
+        self.keyboards.append(keyboard.Keyboard(self, device))
+        self.seat.set_keyboard(device)
+
+    def focus_window(self, win: window.Window, surface: Surface = None):
+        if surface is None:
+            surface = win.surface.surface
+
+        previous_surface = self.seat.keyboard_state.focused_surface
+        if previous_surface == surface:
+            return
+
+        if previous_surface is not None:
+            # Deactivate the previously focused surface
+            previous_xdg_surface = xdg_shell.XdgSurface.from_surface(previous_surface)
+            previous_xdg_surface.set_activated(False)
+
+        # roll the given surface to the front of the list, copy and modify the
+        # list, then save back to prevent any race conditions on list
+        # modification
+        windows = self.windows[:]
+        windows.remove(win)
+        windows.append(win)
+        self.win = windows
+        # activate the new surface
+        win.surface.set_activated(True)
+        self.seat.keyboard_notify_enter(surface, self.seat.keyboard)
+        logger.debug("Focussed new window")
+
+    def setup_listener(self, qtile: Qtile) -> None:
+        """Setup a listener for the given qtile instance"""
+        logger.debug("Adding io watch")
+        self.qtile = qtile
+        self.fd = lib.wl_event_loop_get_fd(self.event_loop._ptr)
+        asyncio.get_running_loop().add_reader(self.fd, self._poll)
+
+    def remove_listener(self) -> None:
+        """Remove the listener from the given event loop"""
+        if self.fd is not None:
+            logger.debug("Removing io watch")
+            loop = asyncio.get_running_loop()
+            loop.remove_reader(self.fd)
+            self.fd = None
+
+    def _poll(self) -> None:
+        self.display.flush_clients()
+        self.event_loop.dispatch(-1)
+
+    def update_desktops(self, groups: List[group._Group], index: int) -> None:
+        """Set the current desktops of the window manager
+
+        The list of desktops is given by the list of groups, with the current
+        desktop given by the index
+        """
+        new_count = len(groups)
+        while new_count > self.desktops:
+            self.desktops += 1
+        while new_count < self.desktops:
+            self.desktops -= 1
+        self.current_desktop = index
+
+    def get_screen_info(self) -> List[Tuple[int, int, int, int]]:
+        """Get the screen information"""
+        xywh = {}  # type: Dict[Tuple[int, int], Tuple[int, int]]
+        for screen in self.outputs:
+            sx, sy, sw, sh = screen.get_geometry()
```

Comment:
It may make sense for the screen info to just return this screen geometry information, and the logic around combining screens can be kept in a single shared location between backends.

Reply:
Yeah good call, updated.